docs/admin/settings.md
# Admin Settings Panel

The GoToSocial admin settings panel uses the [admin API](https://docs.gotosocial.org/en/latest/api/swagger/#operations-tag-admin) to manage your instance. It's combined with the [user settings panel](../user_guide/settings.md) and uses the same OAuth mechanism as normal clients (with scope: admin).

## Setting admin account permissions and logging in

To use the admin settings panel, your account has to be promoted to admin:

```bash
./gotosocial --config-path ./config.yaml admin account promote --username YOUR_USERNAME
```

In order for the promotion to 'take', you may need to restart your instance after running the command.

After this, you can navigate to `https://[your-instance-name.org]/settings`, enter your domain in the login field, and login like you would with any other client. You should now see the admin settings.

## Moderation

Instance moderation settings.

### Reports

![List of reports for testing, showing one open report.](../public/admin-settings-reports.png)

The reports section shows a list of reports, originating from your local users, or remote instances (shown anonymously as just the name of the instance, without specific username).

Clicking a report shows if it was resolved (with the reasoning if available), more information, and a list of reported toots if selected by the reporting user. You can also use this view to mark a report as resolved, and fill in a comment. Whatever comment you enter here will be visible to the user that created the report, if that user is from your instance.

![The detailed view of an open report, showing the reported status and the reason for the report.](../public/admin-settings-report-detail.png)

Clicking on the username of the reported account opens that account in the 'Accounts' view, allowing you to perform moderation actions on it.

### Accounts

You can use this section to search for an account and perform moderation actions on it.

### Domain Permissions

![List of suspended instances, with a field to filter/add new blocks. Below is a link to the bulk import/export interface](../public/admin-settings-federation.png)

In the domain permissions section you can create, delete, and review domain blocks, domain allows, drafts, excludes, and subscriptions.

For more detail on federation settings, and specifically how domain allows and domain blocks work in combination, please see [the federation modes section](./federation_modes.md), and [the domain blocks section](./domain_blocks.md).

#### Domain Blocks

You can enter a domain to suspend in the search field, which will filter the list to show you if you already have a block for it.

Clicking 'suspend' gives you a form to add a public and/or private comment, and submit to add the block.

Adding a domain block will suspend all currently known accounts from that domain, and prevent any new interactions with the blocked domain.

#### Domain Allows

The domain allows section works much like the domain blocks section, described above, only for explicit domain allows rather than domain blocks.

#### Import/export

In this section you can do bulk import/export of domain permissions in JSON, CSV, or plaintext formats.

![List of domains included in an import, providing ways to select some or all of them, change their domains, and update the use of subdomains.](../public/admin-settings-federation-import-export.png)

Upon importing a list, either through the input field or from a file, you can review the entries in the list before importing a subset. You'll also be warned for entries that use subdomains, providing an easy way to change them to the main domain.

#### Drafts

In this section you can create, search through, accept, and reject domain permission drafts.

Domain permission drafts are domain permissions that have been proposed (either via manual creation or as an entry from a subscribed block / allow list), but have not yet come into force.

Until it is accepted, a domain permission draft will not have any effect on federation with the domain it targets. Upon acceptance, it will be converted into either a domain block or a domain allow, and start being enforced.

#### Excludes

In this section, you can create, search through, and remove domain permission excludes.

Domain permission excludes prevent permissions for a domain (and all subdomains) from being automatically managed by domain permission subscriptions.

For example, if you create an exclude entry for the domain `example.org`, then a blocklist or allowlist subscription will exclude entries for `example.org` and any of its subdomains (`sub.example.org`, `another.sub.example.org` etc.) when creating domain permission drafts and domain blocks/allows.

This functionality allows you to manually manage permissions for excluded domains, in cases where you know you definitely do or don't want to federate with a given domain, no matter what entries are contained in a domain permission subscription.

Note that by itself, creation of an exclude entry for a given domain does not affect federation with that domain at all, it is only useful in combination with permission subscriptions.

#### Subscriptions

In this section, you can create, search through, edit, test, and remove domain permission subscriptions.

Domain permission subscriptions allow you to specify a URL at which a permission list is hosted. Every 24hrs at 11pm (by default), your instance will fetch and parse each subscribed list, and create domain permissions (or domain permission drafts) based on entries in the lists.

##### Title

You can optionally use the title field to set a title for the subscription, as a reminder for yourself and other admins.

For example, you might subscribe to a list at `https://lists.example.org/baddies.csv` and set the title of the subscription to something that reflects the contents of that list, such as "Basic block list (worst of the worst)", or similar.

##### Subscription Priority

When you specify multiple domain permission subscriptions, they will be fetched and parsed in order of priority, from highest priority (255) to lowest priority (0).

Permissions discovered on lists higher up in the priority ranking will override permissions on lists lower down in the priority ranking.

For more information on priority, please see the separate [domain permission subscriptions](./domain_permission_subscriptions.md) document.

##### Permission Type

You can use this dropdown to select whether permissions discovered at the list URL should be created as domain blocks, or domain allows.

##### Content Type

You can use this dropdown to select the content type of the list at the subscribed URL.

Use CSV for Mastodon-style permission lists, plain for plaintext lists of domain names, or JSON for json-exported lists.

##### Basic Auth

Check this box to provide a basic auth username and/or password credential for the subscribed list, which will be sent along with each request to fetch the list.

##### Adopt Orphan Permissions

If you check this box, then any existing domain permissions will become managed by this subscription in the following circumstances:

1. They don't already have a subscription ID (ie., they're not managed by any domain permission subscription).
2. They match a domain permission included in the list at the URL of this subscription.

For more information on orphan permissions, please see the separate [domain permission subscriptions](./domain_permission_subscriptions.md#orphan-permissions) document.

##### Remove Retracted Permissions

This setting controls how retractions are handled by this domain permission subscription: if "Remove retracted permissions" is checked, retracted entries will be removed from the database; if "Remove retracted permissions" is not checked, retracted entries will just be orphaned instead.

For more detail on how retractions work, with examples, please see the separate [domain permission subscriptions](./domain_permission_subscriptions.md#retractions) document.

##### Create Permissions as Drafts

With this box checked (default), any permissions created by this subscription will be created as **drafts** which require manual approval to come into force.

It is recommended to leave this box checked unless you absolutely trust the subscribed list, to avoid inadvertent blocking or allowing of domains you'd rather not block or allow.

##### Test a Subscription

To test whether a subscription can be successfully parsed, first create the subscription, then in the detailed view for that subscription, click on the "Test" button.

If your instance is able to fetch and parse permissions at the subscription URI, then you will see a list of these after clicking "Test". Otherwise, you will see an error message.

![Screenshot of the detailed view of a subscription, with arrows pointing to the test section near the bottom.](../public/admin-settings-federation-subscription-test.png)

## Administration

Instance administration settings.

### Actions

Run one-off administrative actions.

#### Email

You can use this section to send a test email to the given email address, with an optional test message.

#### Media

You can use this section run a media action to clean up the remote media cache using the specified number of days. Media older than the given number of days will be removed from storage (s3 or local). Media removed in this way will be refetched again later if the media is required again. This action is functionally identical to the media cleanup that runs automatically.

#### Keys

You can use this section to expire/invalidate public keys from the selected remote instance. The next time your instance receives a signed request using an expired key, it will attempt to fetch and store the public key again.

### Custom Emoji

Custom Emoji will be automatically fetched when included in remote toots, but to use them in your own posts they have to be enabled on your instance.

#### Local

![Local custom emoji section, showing an overview of custom emoji sorted by category. There are a lot of garfields.](../public/admin-settings-emoji-local.png)

This section shows an overview of all the custom emoji enabled on your instance, sorted by their category. Clicking an emoji shows it's details, and provides options to change the category or image, or delete it completely. The shortcode cannot be updated here, you would have to upload it with the new shortcode yourself (and optionally delete the old one).

Below the overview you can upload your own custom emoji, after previewing how they look in a toot. PNG and (animated) GIF's are supported.

#### Remote

![Remote custom emoji section, showing a list of 3 emoji parsed from the entered toot, garfield, blobfoxbox and blobhajmlem. They can be selected, their shortcode can be tweaked, and they can be assigned to a category, before submitting as a copy or delete operation](../public/admin-settings-emoji-remote.png)

Through the 'remote' section, you can look up a link to any remote toots (provided the instance isn't suspended). If they use any custom emoji they will be listed, providing an easy way to copy them to the local emoji (for use in your own toots), or disable them ( hiding them from toots).

**Note:** as the testrig server does not federate, this feature can't be used in development (500: Internal Server Error).

### Instance Settings

![Screenshot of the GoToSocial admin panel, showing the fields to change an instance's settings](../public/admin-settings-instance.png)

Here you can set various metadata for your instance, like the displayed name/title, thumbnail image, (short) description, and contact info.

#### Instance Appearance

These settings primary affect how your instance appears to others and on the web.

Your **instance title** will appear at the top of every web page on your instance, and in OpenGraph meta tags, so pick something that represents the vibe of your instance.

The **instance avatar** is sort of like the mascot of your instance. It will appear next to the instance title at the top of every page, and as the preview image in browser tabs, OpenGraph links, and that sort of thing.

If you set an instance avatar, we highly recommend setting the **avatar image description** as well. This will provide alt text for the image you set as avatar, helping screenreader users to understand what's depicted in the image. Keep it short and sweet.

#### Instance Descriptors

You can use these fields to set short and full descriptions of your instance, as well as to provide terms and conditions for current and prospective users of your instance.

The **short description** will be shown on the instance home page, right near the top, and in response to `/api/v1/instance` queries.

It's a good idea to provide something pithy in here, to give visitors to your instance an immediate impression of what you're all about. For example:

> This is an instance for enthusiasts of classic synthesizers.
>
> Sick beats are for life, and not just for Christmas!

or:

> This is a single-user instance just for me!
>
> Here's my profile: @your_username

The **full description** will appear on your instance's /about page, and in response to `/api/v1/instance` queries.

You can use this to provide info like:

- your instance's history, ethos, attitude, and vibe
- the kinds of things your instance denizens tend to post about
- how to get an account on your instance (if it's possible at all)
- a list of users with accounts on the instance, who want to be found more easily

The **terms and conditions** box also appears on your instance's /about page, and in response to `/api/v1/instance` queries.

Use it for filling in stuff like:

- legal jargon (imprint, GDPR, or links thereto)
- federation policy
- data policy
- account deletion/suspension policy

All of the above fields accept **markdown** input, so you can write proper lists, codeblocks, horizontal rules, block quotes, or whatever you like.

You can also mention accounts using the standard `@user[@domain]` format.

Have a look at the [markdown cheat sheet](https://markdownguide.offshoot.io/cheat-sheet/) to see what else you can do.

### Instance Contact Info

In this section, you can provide visitors to your instance with a convenient way of reaching your instance admin.

Links to the set contact account and/or email address will appear on the footer of every web page of your instance, on the /about page in the "contact" section, and in response to `/api/v1/instance` queries.

The selected **contact user** must be an active (not suspended) admin and/or moderator on the instance.

If you're on a single-user instance and you give admin privileges to your main account, you can just fill in your own username here; you don't need to make a separate admin account just for this.

### Instance Custom CSS

custom CSS allows you to further customize the way your instance looks when visited through a browser.

This custom CSS will be applied to all pages of your instance. Users themes and CSS still take precedence over this customization.

See the [Custom CSS](../user_guide/custom_css.md) page for some tips on writing custom CSS for your instance.


activity/CHANGELOG
v1.0.0 2020-07-09

* Rename Callbacks to FederatingCallbacks in FederatingProtocol.
* Rename Callbacks to SocialCallbacks in SocialProtocol.
* Rename NewID to NewID on the Database interface.
* Added approx. 200 unit tests at ~70% LOC coverage.
* More fixes where nil pointers would be dereferenced.
* Modified programmatic delivery to be lenient instead of strict.
* Even more 'astool' bugfixes to handle multiple vocabulary code generation.
* Added request lifecycle hooks to allow implementations greater control when
      using the 'context' standard library package and its 'Context'.
* Fixed bugs in programmatic delivery, handling rare forms of HTTP successes,
      handling of JSON-LD's @context, race conditions.
* Provide a Mastodon-compliant http-signatures Transport implementation, when
      using the github.com/go-fed/httpsig library.
* Support programmatic sending of Activities
* Added support for 3 vocabularies in addition to ActivityStreams: toot
      (Mastodon), security v1, and ForgeFed.
* Modify 'astool' to address quirky logic and underspecification in various
      vocabulary definitions.
* Fixed bugs in the 'astool' logic.
* Added more helpful functions to the code generated by 'astool'.
* Inbox forwarding bug fixes.
* Migrate to go modules.

2019-02-24

* Removed 'tools'.
* Removed 'vocab'.
* Removed 'deliver'.
* Created the new 'astool' for code-generating any ActivityStreams vocabulary.
* The 'streams' package has entirely been redesigned and regenerated.
* The 'pub' package has been redesigned to be extensible, and has had concepts
      previously abstracted, such as security, removed in favor of opaque
      methods that are up to the application to implement.
* This succinct summary betrays the size, scope, and effort into rethinking
      this ActivityPub library.

v0.4.0 2018-11-17

* The 'streams' package now has constructors for each of its generated types.
* The README now lists WriteFreely instead of Write.as as a user of this
      library, though Write.as does also use go-fed. WriteFreely is the open-
      source fork of the proprietary Write.as.
* The 'tools/exp' subdirectory contains a half-built work-in-progress tool that
      will be used to generate the v1.X versions of this library. It does not
      materially affect the v0.X versions' functionality nor its code
      generation. Its presence is merely a consequence of the go-fed/activity
      maintainer [continuing to] fail at using git branches.

v0.3.0 2018-08-21

* Interfaces in 'vocab' now properly contain all of the unknown methods
      available for all properties. These methods were available on the structs
      themselves but only partially listed in the interfaces. Specifically,
      multiply-typed properties (regardless of being functional or not) were
      omitted.
* 'pub' now supports Announce activities' default behavior. Upon receiving an
      Announce activity via S2S, it is added to the 'shares' property of
      object(s) owned in the application.
* Add the missing 'shares' property to 'vocab.Object' and its child types as
      well as 'streams.Object' and its child types. It is an ActivityPub
      specific property not a part of the ActivityStreams specification.
* The Callbackers for the SocialAPI and FederateAPI can have additional
      methods of the form 'X(c context.Context, s *streams.X) error' where X is
      a Core or Extended type not already a part of the Callbacker interface.
      This lets client applications decide which Activity types need further
      handling, without being burdened of implementing unused stub methods. The
      new Activities supported have no default behavior supported by 'pub',
      except for the Announce activity as desribed above. An exhaustive list of
      Activities that can be X (as mentioned above) are:
          - Announce
          - Arrive
          - Dislike
          - Flag
          - Ignore
          - Invite
          - Join
          - Leave
          - Listen
          - Move
          - Offer
          - Question
          - Read
          - TentativeAccept
          - TentativeReject
          - Travel
          - View
* The 'vocab' package now properly generates Intransitive Activity subtype
      structs such that they are now convertible to the
      vocab.IntransitiveActivityType interface.
* Types in the 'vocab' package now support setting, getting, removing, and
      testing existence for unknown properties (extended properties).

v0.2.1 2018-08-19

* Request body is now correctly copied when sending federation messages.
* Change RWType and FollowResponse to bool and uint8, respectively.
* Update README with applications using the go-fed/activity library.
* Update README with links to official implementation reports for go-fed.
* Add unit test to document behavior when maxDeliveryDepth is set to zero.

v0.2.0 2018-08-04

* Begin FederateAPI unofficial implementation report.
* All 'vocab.Object' types and types extending from 'vocab.Object' now have an
      'IsPublic' method that will return true if the 'to', 'bto', 'cc', or 'bcc'
      properties have the ActivityPub special Public collection IRI. The
      'streams' types also have a corresponding 'IsPublic' method.
* Use 'OrderedCollection' as the default type for 'likes', 'liked',
      'following', and 'followers' properties if the actor or object does not
      have an IRI, 'Collection', or 'OrderedCollection' set for these
      properties.
* Examine the IRI of 'objects' when applying the Origin Check policy for Update
      and Delete activities.
* If a federated Activity was already received, do not execute its side effects
      a second time.
* Add 'Like' activities to the 'likes' collection, instead of adding the
      actors. This was a specification-violating behavior.
* No longer try to fetch IRIs when deduping by IRI.
* Remove unused methods from fed_test.go.
* Fix Media Type header detection for ActivityPub messages.

testrig/config.go
// GoToSocial
// Copyright (C) GoToSocial Authors admin@gotosocial.org
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package testrig

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"time"

	"code.superseriousbusiness.org/gotosocial/internal/config"
	"code.superseriousbusiness.org/gotosocial/internal/language"
	"code.superseriousbusiness.org/gotosocial/internal/media/ffmpeg"
	"codeberg.org/gruf/go-bytesize"
	"github.com/coreos/go-oidc/v3/oidc"
)

func init() {
	ctx := context.Background()

	// Ensure global ffmpeg WASM pool initialized.
	fmt.Println("testrig: precompiling ffmpeg WASM")
	if err := ffmpeg.InitFfmpeg(ctx, 1); err != nil {
		panic(err)
	}

	// Ensure global ffmpeg WASM pool initialized.
	fmt.Println("testrig: precompiling ffprobe WASM")
	if err := ffmpeg.InitFfprobe(ctx, 1); err != nil {
		panic(err)
	}
}

// InitTestConfig initializes viper
// configuration with test defaults.
func InitTestConfig() {
	config.Defaults = testDefaults()
	config.Reset()
}

func testDefaults() config.Configuration {
	return config.Configuration{
		LogLevel:                   envStr("GTS_LOG_LEVEL", "error"),
		LogTimestampFormat:         "02/01/2006 15:04:05.000",
		LogDbQueries:               true,
		ApplicationName:            "gotosocial",
		LandingPageUser:            "",
		ConfigPath:                 "",
		Host:                       "localhost:8080",
		AccountDomain:              "localhost:8080",
		Protocol:                   "http",
		BindAddress:                "127.0.0.1",
		Port:                       8080,
		TrustedProxies:             []string{"127.0.0.1/32", "::1"},
		DbType:                     envStr("GTS_DB_TYPE", "sqlite"),
		DbAddress:                  envStr("GTS_DB_ADDRESS", ":memory:"),
		DbPort:                     envInt("GTS_DB_PORT", 0),
		DbUser:                     envStr("GTS_DB_USER", ""),
		DbPassword:                 envStr("GTS_DB_PASSWORD", ""),
		DbDatabase:                 envStr("GTS_DB_DATABASE", ""),
		DbTLSMode:                  envStr("GTS_DB_TLS_MODE", ""),
		DbTLSCACert:                envStr("GTS_DB_TLS_CA_CERT", ""),
		DbPostgresConnectionString: envStr("GTS_DB_POSTGRES_CONNECTION_STRING", ""),
		DbMaxOpenConnsMultiplier:   8,
		DbSqliteJournalMode:        "WAL",
		DbSqliteSynchronous:        "NORMAL",
		DbSqliteCacheSize:          8 * bytesize.MiB,
		DbSqliteBusyTimeout:        time.Minute * 5,

		WebTemplateBaseDir: "./web/template/",
		WebAssetBaseDir:    "./web/assets/",

		InstanceFederationMode:         config.InstanceFederationModeDefault,
		InstanceFederationSpamFilter:   true,
		InstanceExposePeers:            true,
		InstanceExposeBlocklist:        true,
		InstanceExposeBlocklistWeb:     true,
		InstanceExposeAllowlist:        true,
		InstanceExposeAllowlistWeb:     true,
		InstanceExposeCustomEmojis:     true,
		InstanceDeliverToSharedInboxes: true,
		InstanceLanguages: language.Languages{
			{
				TagStr: "nl",
			},
			{
				TagStr: "en-gb",
			},
		},
		InstanceSubscriptionsProcessFrom:  "23:00",        // 11pm,
		InstanceSubscriptionsProcessEvery: 24 * time.Hour, // 1/day.
		InstanceAllowBackdatingStatuses:   true,

		AccountsRegistrationOpen:         true,
		AccountsReasonRequired:           true,
		AccountsRegistrationDailyLimit:   10,
		AccountsRegistrationBacklogLimit: 20,
		AccountsAllowCustomCSS:           true,
		AccountsCustomCSSLength:          10000,
		AccountsMaxProfileFields:         8,

		Media: config.MediaConfiguration{
			DescriptionMinChars: 0,
			DescriptionMaxChars: 500,
			RemoteCacheDays:     7,
			LocalMaxSize:        40 * bytesize.MiB,
			RemoteMaxSize:       40 * bytesize.MiB,
			EmojiLocalMaxSize:   51200,          // 50KiB
			EmojiRemoteMaxSize:  102400,         // 100KiB
			CleanupFrom:         "00:00",        // midnight.
			CleanupEvery:        24 * time.Hour, // 1/day.
			ThumbMaxPixels:      512,
		},

		// the testrig only uses in-memory storage, so we can
		// safely set this value to 'test' to avoid running storage
		// migrations, and other silly things like that
		StorageBackend:       "test",
		StorageLocalBasePath: "",

		StatusesMaxChars:           5000,
		StatusesPollMaxOptions:     6,
		StatusesPollOptionMaxChars: 50,
		StatusesMediaMaxFiles:      6,

		LetsEncryptEnabled:      false,
		LetsEncryptPort:         0,
		LetsEncryptCertDir:      "",
		LetsEncryptEmailAddress: "",

		OIDCEnabled:          false,
		OIDCIdpName:          "",
		OIDCSkipVerification: false,
		OIDCIssuer:           "",
		OIDCClientID:         "",
		OIDCClientSecret:     "",
		OIDCScopes:           []string{oidc.ScopeOpenID, "profile", "email", "groups"},
		OIDCLinkExisting:     false,
		OIDCAdminGroups:      []string{"adminRole"},
		OIDCAllowedGroups:    []string{"allowedRole"},

		SMTPHost:               "",
		SMTPPort:               0,
		SMTPUsername:           "",
		SMTPPassword:           "",
		SMTPFrom:               "GoToSocial",
		SMTPDiscloseRecipients: false,

		TracingEnabled: false,
		MetricsEnabled: false,

		SyslogEnabled:  false,
		SyslogProtocol: "udp",
		SyslogAddress:  "localhost:514",

		Advanced: config.AdvancedConfig{
			CookiesSamesite:  "lax",
			SenderMultiplier: 0, // 1 sender only, regardless of CPU

			RateLimit: config.RateLimitConfig{
				Requests: 0, // disabled
			},

			Throttling: config.ThrottlingConfig{
				Multiplier: 0, // disabled
			},

			ScraperDeterrence: config.ScraperDeterrenceConfig{
				Enabled:    envBool("GTS_ADVANCED_SCRAPER_DETERRENCE_ENABLED", false),
				Difficulty: uint32(envInt("GTS_ADVANCED_SCRAPER_DETERRENCE_DIFFICULTY", 100000)), //nolint
			},
		},

		SoftwareVersion: "0.0.0-testrig",

		// simply use cache defaults.
		Cache: config.Defaults.Cache,
	}
}

func envBool(key string, _default bool) bool {
	return env(key, _default, func(value string) bool {
		b, _ := strconv.ParseBool(value)
		return b
	})
}

func envInt(key string, _default int) int {
	return env(key, _default, func(value string) int {
		i, _ := strconv.Atoi(value)
		return i
	})
}

func envStr(key string, _default string) string {
	return env(key, _default, func(value string) string {
		return value
	})
}

func env[T any](key string, _default T, parse func(string) T) T {
	value, ok := os.LookupEnv(key)
	if ok {
		return parse(value)
	}
	return _default
}
